<h1><b>Statistique en Bioinformatique : </b> TME7</h1>
<br>
L’objectif de ce TME est: 
<br>
<ul>
<li> Construire un modèle d'une famille proteique avec HMM profil. Ce modèle facilite la reconnaissance des membres de la famille (temps de calcul, calibration du seuil de significativité), en particulier pour les protéines les plus distantes (point de vue évolutif). </li> 
</ul>
<br>
<div class="alert alert-warning" role="alert" style="margin: 10px">
<p>**Soumission**</p>
<ul>
<li>Renomer le fichier TME8.ipynb pour NomEtudiant1_NomEtudiant2_TME8.ipynb </li>
<li>Envoyer par email à edoardo.sarti@upmc.fr, l’objet du email sera [SBAS-2019] TME8</li>
</ul>
</div>


Nom etudiant 1 :
Yann Zhong

Nom etudiant 2 :
Alexis Trang

<h3> Installer HMMer</h3>
Nous allons travailler avec deux versions du logiciel  HMMer: 2.3.2 et 3. Installez les deux versions dans repertoires differents. Tout d'abord, nous allons considerer la version 2.3.2<br><br>

<h4>-Installer la version 2.3.2</h4><br>
Télécharger HMMer <a href="http://eddylab.org/software/hmmer/2.3/hmmer-2.3.tar.gz">version 2.3.2</a> <br>
Lisez les instructions et installez-le dans votre espace personnel. <br><br>

<font face="monospace">tar -xzf hmmer-2.3.tar.gz</font><br>
<font face="monospace">cd hmmer-2.3</font><br>
<font face="monospace">./configure --prefix=/home/yourUser/yourDirectory/hmmer-2.3</font><br>
<font face="monospace">make</font><br>

<font color="red"> Remplacez yourUser/yourDirectory par votre coordonnée personnelle.</font>

<h4>-Dans le terminal modifier le fichier .bashrc</h4> 

<font face="monospace">cd \$HOME </font><br>
<font face="monospace">gedit .bashrc </font><br>
Ajouter la ligne suivante à votre fichier <font face="monospace">.bashrc</font><br>
<font face="monospace">export PATH=/home/yourUser/yourDirectory/hmmer-2.3/src:${PATH}</font><br>

<font color="red"> Remplacez <font face="monospace">yourUser/yourDirectory</font> par votre coordonnée personnelle.</font><br>
Sauvergardez le fichier et taper dans le terminal <br>

<font face="monospace">source .bashrc </font><br>




<b>Exercice 1 </b>: <u>Zinc finger Pfam domain</u> <br>
<b>1.1</b>: Construire le modèle HMMer pour le domaine Pfam Zinc finger, numéro d'accès PF00096. Pour cela, téléchargez l'alignement seed au format Stockholm depuis le site web de <a href="http://pfam.xfam.org/family/PF00096#tabview=tab3">Pfam</a>. Construire le modèle HMMer avec les paramètres par défaut, enregistrer avec le nom PF00096.hmm

<b>1.2</b>: Utilisez le programme HMMeditor pour visualiser le modele PF00096.hmm et répondre aux questions.<br>
<b>a)</b>: Combien d'états Match observez-vous?

<font color="blue">
Reponse: On observe 26 états matchs.
</font>

<b>b)</b>: Pourquoi les états M3, M8, M12, M18, M21 et M26 ont une bordure en gras?

<font color="blue">
Reponse: Car la probabilité de transition vers le prochain état match (M4, M9, M13, M22 et END) est de 1 pour tous ces états match. (même si elle est pas exactement 1, car on peut considérer que le rajout d'un pseudo count fait que les probas de transition vers les insert/deletions sont pas nulles)
</font> 



<b>c)</b>: Quelle est la probabilité de transition M5-> I5? 

<font color="blue">
Reponse: 0.0577374030160831
</font>

<b>d)</b>: Quel est l'acide aminé le plus fréquent dans l'état I5?



<font color="blue">
Reponse: L'acide aminé G, avec un taux d'émission de 0.09.
</font>

<b>1.3</b>: Télécharger la séquence X6NX52 à <a href="http://www.uniprot.org/uniprot/X6NX52">Uniprot</a>. Utilisez le programme <font face="monospace">hmmsearch</font> pour localiser les domaines  Zinc finger dans la séquence X6NX52. Combien de domaines obtenez-vous? Comparez vos résultats avec les résultats de  <a href="http://pfam.xfam.org/protein/X6NX52_RETFI">Pfam</a>. Avez-vous trouvé les mêmes domaines dans les mêmes positions? Expliquez votre réponse.

<font color="blue">
Reponse: Les deux retournent quatre domaines, mais Pfam indique que seulement deux de ces domaines sont connus (zf-C2H2 et zf-met). Hmmsearch donne exactement le même domaine que Pfam pour le premier (allant de 74 à 97), mais le deuxième est un peu différent (221 - 242 Pfam, 221 - 244 hmmsearch). Les deux derniers domaines, quant à eux, sont très différents. Ceci s'explique probablement par le fait que hmmsearch applique un seuil différent à celui de Pfam.
</font>

<b>1.4</b>: Enregistrer la séquence du premier domaine, positions 74-97, dans un format fasta avec le nom de fichier dom1.fasta, supprimer les gaps de la séquence avant de l'enregistrer. Chargez le fichier dom1.fasta dans le programme HMMeditor et cliquez sur "Align Path". Qu'observez-vous? 

<font color="blue">
Reponse: On a mit les deux lignes ci-dessus dans dom1.fasta. On observe qu'un "chemin" en bleu est montré allant de matchs vers inserts/deletions, qui vraisemblablement va indiquer le chemin le plus probable pour notre séquence en input donnée.
</font>

<b>1.5</b>: Modifiez le fichier dom1.fasta et ajouter le mot RWV après le mot TFR. Rechargez le fichier dans le programme HMMeditor et clique sur "Align Path". Qu'observez vous?

<font color="blue">
Reponse: Le chemin est légèrement modifié: au lieu d'aller de M15 à M16 (W à G) on passe par I15 (V V R en insert) pour ensuite continuer notre chemin vers M16. On remarque néanmoins que la probabilité d'aller de M15 à M16 reste de 0.9997416301892784 et que celle d'aller de M15 à I15 est minime.
</font>

<h4>-Installer la version 3.</h4><br>
Télécharger HMMer <a href="http://eddylab.org/software/hmmer3/3.1b2/hmmer-3.1b2.tar.gz">version 3</a> <br>
Lisez les instructions et installez-le dans votre espace personnel. <br>

<font face="monospace">tar -xzf hmmer-3.1b2.tar.gz</font><br>
<font face="monospace">cd hmmer-3.1b2</font><br>
<font face="monospace">./configure --prefix=/home/yourUser/yourDirectory/hmmer-3.1b2</font><br>
<font face="monospace">make</font><br>

<font color="red"> Remplacer yourUser/yourDirectory avec votre donnée personnelle.</font>

<h4>-Dans le terminal modifier le fichier .bashrc</h4>

<font face="monospace">cd \$HOME </font><br>
<font face="monospace">gedit .bashrc </font><br>
Mettre en commentaire le export de hmmer-2.3 (utiliser # au début de la ligne) et ajouter la ligne suivante à votre fichier .bashrc<br>
<font face="monospace">export PATH=$PATH:/home/yourUser/yourDirectory/hmmer-3.1b2/src;</font>
<font color="red"> Remplacer yourUser/yourDirectory avec votre donnée personnelle.</font>
Sauvegarder le fichier .bashrc et dans le terminal taper<br>
<font face="monospace">source .bashrc </font><br>

<b>Exercice 2 </b>: <u>Ubiquitin domain Pfam domain</u> <br>
<b>2.1</b>: Construire le modèle HMMer pour le domaine Ubiquitin Pfam, accession number PF00240. Pour cela, téléchargez l'alignement seed au format Stockholm depuis le site web de  <a href="http://pfam.xfam.org/family/PF00240#tabview=tab3">Pfam</a>. Construire le modèle hmmer avec les paramètres par défaut, enregistrer sous le nom PF00240.hmm. Combien d'états Match observez-vous?


<font color="blue">
Reponse: N.B: il faut d'abord convertir le fichier hmm en version 2 pour pouvoir le visualiser. Ici, on observe 72 états match.
</font>

<b>2.2</b>:
Téléchargez toutes les séquences avec le domaine ubiquitin du site Pfam. Clique sur l'alignement, sur Format an alignment, sélectionnez Full, Format FASTA,  Gaps : No gaps unaligned. Enregistrez le fichier sous le nom PF00240_full.fasta. Combien de séquences observez-vous dans le fichier PF00240_full.fasta?

In [24]:
#Code en python ou en bash pour compter les sequences
PF00240 = open("PF00240_full.fasta","r")
seq_count = 0

for line in PF00240:
    if line[0] == '>':
        seq_count += 1
print("Number of sequences:",seq_count)

Number of sequences: 47633


<font color="blue">
Reponse: On observe 47633 séquences en comptant tous les ">".
</font>

<b>2.3</b>:
Utilisez hmmsearch (default parameters) pour détecter les domaines ubiquitin dans le fichier PF00240_full.fasta. Utilisez l'option --domtblout pour enregistrer un fichier tabulaire, donnez le nom PF00240.domtb, répondre aux questions suivantes:

In [7]:
#Read the domtblout file
def read_domtblout(input_f):
    seq_count = 0
    target = open(input_f,"r")
    
    for line in target:
        if line[0] != "#":
            seq_count += 1
    
    return seq_count

read_domtblout("PF00240.domtb")

47657

<b>a)</b>: Combien de domaines observez-vous dans le fichier PF00240.domtb? Comparez le nombre de domaines détectés avec le nombre attendu (question 2.2)

<font color="blue">
Reponse: On en observe 47657, comparé à 47633 par rapport à ce qu'on a compté dans le FASTA.
</font>

<b>b)</b>: Quelle est le c-evaleus et les scores du meilleur et du pire hit?

In [4]:
def best_and_worst(file_name):
    file=open(file_name,'r')
    flag=0
    for line in file:
        if line[0]!='#':
            line=line.strip().split()
            if flag==0:
                min_score=max_score=float(line[7])
                min_c_eval=max_c_eval=float(line[11])
                flag=1
            if min_score>float(line[7]):
                min_score=float(line[7])
                min_c_eval=float(line[11])
            if max_score<float(line[7]):
                max_score=float(line[7])
                max_c_eval=float(line[11])
    return min_score ,min_c_eval ,max_score ,max_c_eval

In [5]:
best_and_worst("PF00240.domtb")

(24.1, 6.3e-05, 120.9, 4e-35)

<font color="blue">
Reponse: en considérant que le meilleur hit est le tout premier et que le pire est le tout dernier, on aura pour meilleur hit R0KPY3_NOSB1/3-74 avec un score de 120.9 et un e-value de 3.7e-35, et le pire hit A0A1U7VJ77_NICSY/1-50 avec un score de 24.1 et un e-value de 6e-05.
</font>

<b>c)</b>: Exécutez à nouveau hmmsearch et définissez le domain e-value à 1e-5. Combien de domaines observez-vous? Comparez le nombre de domaines détectés avec l'attendu (question 2.2) et obtenu dans la question a. Justifiez votre réponse.

In [9]:
read_domtblout("PF00240_2.domtb")

47271

<font color="blue">
Reponse: On observe 47271 domaines, donc le fait de donner un seuil au e-value nous a retourné 386 domaines en moins avec hmmsearch. On garde alors tous les domaines identifiés avec un e-value en dessous de 1e-5. En regardant et en comparant les deux fichiers domtb, on voit que à partir de ligne 47540, les résultats obtenus avec le premier hmmsearch sont exclus (à partir de A0A4D9AK69_SALSN/159-232).
</font>

<b>d)</b>:
Téléchargez le modèle hmm pour le domaine ubiquitin depuis le site de Pfam. Clique sur "Curation & model" et sur "download the raw HMM for this family". Exécutez à nouveau hmmsearch avec l'option --cut_ga. Comparez le nombre de domaines détectés avec les questions précédentes. Qu'est-ce que cut_ga fait?

In [11]:
read_domtblout("PF00240_3.domtb")

47625

<font color="blue">
Reponse: L'option cut_ga (qui est incompatible avec l'option domE) permet d'utiliser le seuil Pfam GA (gathering threshold score cutoffs). Dans ce cas ci on trouve 47625 domaines, qui est plus que avec le seuil de 1e-5 mais toujours moins que sans option.
</font>

<b>Exercice 3 </b>: <u>Cooccurrence de domaines</u> <br>
Le domaine Ribosomal_S27 (PF01599) est très suivant trouvé avec le domaine de l'ubiquitine?  Allez sur le site Pfam, recherchez le domaine PF01599 et cliquez sur "Domain Organisation" recherchez le domaine ubiquitin dans la liste des architectures de domaine. Répondre aux question suivantes:<br>
<b>3.1</b>:
Si ces domaines coexistent, dans combien de séquences cette architecture de domaines est-elle observée? 

<font color="blue">
Reponse: On observe 1547 séquence avec une architecture Ubiquitine+Ribosomal_S27, de plus on trouve 8 avec une architecture 2*Ubiquitine+Ribosomal_S27.
</font>

<b>3.2</b>: Vérifions si la séquence W4GKD7 décrite comme protéine ribosomale S27a de l'Ubiquitine-40S a l'architecture du domaine: ubiquitine, Ribosomal_S27. <br>
- Téléchargez la séquence de W4GKD7 de <a href="http://www.uniprot.org/uniprot/W4GKD7">Uniprot</a>  et téléchargez le modèle Ribosomal_S27 hmmer à partir du site Web Pfam.
- Concattez les modèles ubiquitin.hmm et Ribosomal_S27.hmm, enregistrez le fichier sous Ribosomal_S27_ubiquitin.hmm.
- Exécutez hmmsearch avec l'option --cut_ga pour rechercher des domaines dans W4GKD7, utilisez le modèle de fichier Ribosomal_S27_ubiquitin.hmm. Enregistrez la sortie au format tabulaire, utilisez l'option --domtblout pour enregistrer un fichier tabulaire, nommé W4GKD7.domtb
- Analyser le fichier W4GKD7.domtb, avez-vous détecté les domaines Ribosomal_S27 et ubiquitin? Quels sont les scores et les e-values?

<font color="blue">
Reponse: on a bien détecté les deux domaines. Pour ribosomal S27, on a un score de 91.2 et e-value de 1.5e-30, et pour ubiquitin, on a un score de 113.0 et e-value de 2.3e-37.
</font>

<b>3.3</b>: Visualisez l'architecture du domaine pour la protéine W4GKD7 en utilisant le programme  <a href="http://prosite.expasy.org/cgi-bin/prosite/mydomains/">myDomains</a>; Remplissez la longueur de la protéine et les données de domaine. Vous pouvez trouver toutes les informations dans le fichier W4GKD7.domtb, choisir la forme et la couleur que vous préférez, inclure l'image dans votre reponse.

![Reponse](files/architecture_W4GKD7.png)

<b>Exercice 4 </b>: <u>Domaines du même clan</u><br>
Les familles de Pfam ayant des fonctions similaires sont suivant orgnisés dans un clan, c'est le cas des familles ubiquitin et ubiquitin_2 placé dans le clan CL0072.<br>
<b>4.1</b>: Combien de membre exists dans le clans CL0072?

<font color="blue">
Reponse: En allant au lien à https://pfam.xfam.org/clan/CL0072 on trouve que le clan CL0072 possède 77 membres (et 309820 domaines associés).
</font>

<b>4.2</b>: Téléchargez le modèle hmm pour le domaine ubiquitin_2 depuis le site de Pfam. Clique sur "Curation & model" et sur "download the raw HMM for this family". Concatenez les modeles ubiquitin.hmm et ubiquitin_2.hmm and executez hmmsearch avec --cut_ga sur les sequences du fichier Uniprot_ubiquitin.fasta, utilizez l'option --domtblout pour enregistrer un fichier tabulaire. Repondrez aux questions:


<b>a)</b>: Combien de sequences ont plus qu'un hit?

<font color="blue">
Reponse: En comptant par rapport aux noms uniques sur le 1e collone, on a 73 séquences, et on note que aucune séquence identifiée n'a qu'une hit.
</font>

<b>b)</b>: Faite une fonction pour filtrer les meilleurs hits dans les séquences de Uniprot_ubiquitin.fasta. Si une séquence a plus qu'un hit dans de positions chevauchantes vous devez choisir le hit avec le meilleur c-evalue. Le début et fin de chaque hit se trouve dans les colonnes 20 et 21 du fichier de sortie de hmmsearch.